slack/slack.go

```diff
@@ -0,0 +1,102 @@
+// Package slack provides an slash Handler for adding the Conveyor push webhook
+// on the GitHub repo.
+package slack
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+
+	"github.com/ejholmes/slash"
+	"github.com/google/go-github/github"
+)
+
+// NewHook returns a new github.Hook instance that represents the appropriate
+// configuration for the Conveyor webhook.
+func NewHook(url, secret string) *github.Hook {
+	return &github.Hook{
+		Events: []string{"push"},
+		Active: github.Bool(true),
+		Name:   github.String("web"),
+		Config: map[string]interface{}{
+			"url":          url,
+			"content_type": "application/json",
+			"secret":       secret,
+		},
+	}
+}
+
+// hooker is something that can add the conveyor webhook to a github repo.
+type hooker interface {
+	CreateHook(owner, repo string, hook *github.Hook) (*github.Hook, *github.Response, error)
+	ListHooks(owner, repo string, opt *github.ListOptions) ([]github.Hook, *github.Response, error)
+	EditHook(owner, repo string, id int, hook *github.Hook) (*github.Hook, *github.Response, error)
+}
+
+// WebhookHandler implements the slash.Handler interface for setting up the
+// conveyor webhook.
+type WebhookHandler struct {
+	Hook *github.Hook
+	hooker
+}
+
+// NewWebhookHandler initializes a new WebhookHandler instance.
+func NewWebhookHandler(c *github.Client, hook *github.Hook) *WebhookHandler {
+	return &WebhookHandler{
+		Hook:   hook,
+		hooker: c.Repositories,
+	}
+}
+
+func (h *WebhookHandler) ServeCommand(ctx context.Context, command slash.Command) (reply string, err error) {
+	params := slash.Params(ctx)
+	owner, repo := params["owner"], params["repo"]
+
+	var hook *github.Hook
+	hook, err = h.existingHook(owner, repo)
+	if err != nil {
+		return
```

Comment:
I almost always prefer to not use named params, but there's a couple of instances where I use them:

1. When you need to reference the return value in a defer.
2. When you're building up a non nilable value, like a string here. Gets annoying to `return "", nil`.

No. 2 has arguable benefits. Just avoids having to sprinkle empty strings in a lot of returns.